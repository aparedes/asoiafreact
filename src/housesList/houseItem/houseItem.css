.house {
  background-color: lightgray;
  border: 1px solid darkgray;
  color: black;
  padding: 10px;
  display: flex;
  flex-direction: column;
}

.houseName {
  padding: 5px;
  font-weight: bold;
}

.houseDetails {
  padding: 5px;
  text-align: left;
}

.houseOpen {
  margin-left: auto;
  margin-right: auto;
  width: 50px;
  height: 50px;
  line-height: 50px;
  text-align: center;
}
