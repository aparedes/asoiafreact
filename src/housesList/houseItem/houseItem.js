/* @flow */
import './houseItem.css'
import * as Immutable from 'immutable'
import React, { Component, PropTypes } from 'react'
import type { Map } from 'immutable'
import { connect } from 'react-redux'

type PropsType = {
  coatOfArms: string,
  error: ?string,
  getHouse: () => void,
  getting: boolean,
  got: boolean,
  houseId: string,
  name: string,
  words: string,
  openHouse: () => void,
}

type StateType = {
  showFull: boolean
}

export class HouseItem extends Component {
  static displayName = 'House'
  static propTypes = {
    coatOfArms: PropTypes.string,
    houseId: PropTypes.string.isRequired,
    name: PropTypes.string,
    words: PropTypes.string,
  }
  changeState: () => void
  openExternal: (evt: TouchEvent) => void
  props: PropsType
  state: StateType
  constructor(props: PropsType) {
    super(props)
    this.changeState = this.changeState.bind(this)
    this.openExternal = this.openExternal.bind(this)
  }
  componentWillMount() {
    this.state = {
      showFull: false,
    }
  }

  componentDidUpdate(prevProps: PropsType, prevState: StateType) {
    if (!prevState.showFull && this.state.showFull && !this.props.got && !this.props.getting) {
      this.props.getHouse()
    }
  }


  changeState() {
    this.setState({
      showFull: !this.state.showFull,
    })
  }

  openExternal(evt: TouchEvent) {
    evt.preventDefault()
    evt.stopPropagation()
    this.props.openHouse()
  }
  renderDetails() {
    const { coatOfArms, words } = this.props
    return (
      <div className={ 'houseDetails' }>
        <div className={ 'houseDetail' }>
          <b>{ 'Coat of Arms: '}</b>{ coatOfArms.length > 0 ? coatOfArms : '?'}
        </div>
        <div className={ 'houseDetail' }>
          <b>{ 'Words: '}</b>{ words.length > 0 ? words : '?' }
        </div>
        <div className={ 'houseOpen' } onClick={ this.openExternal }>
          <i className={ 'fa fa-external-link' } aria-hidden={ 'true' }></i>
        </div>
      </div>
    )
  }
  render() {
    const { error, getting } = this.props
    return (
      <div className={ 'house' } onClick={ this.changeState }>
        <div className={ 'houseName' }>{ this.props.name }</div>
        { this.state.showFull ? getting ? 'Loading' : error ? error : this.renderDetails() : null }
      </div>
    )
  }
}

const mapStateToProps = (state: { houses: Map<string, Map<string, *>> }, { houseId }: { houseId: string }): Object => {
  const house = state.houses.getIn([ 'houses', houseId ], new Immutable.Map())
  return ({
    coatOfArms: house.getIn([ 'coatOfArms' ], ''),
    error: house.getIn([ 'error' ], null),
    getting: house.getIn([ 'getting' ], false),
    got: house.has('coatOfArms'),
    name: house.getIn([ 'name' ], ''),
    words: house.getIn([ 'words' ], ''),
  })
}
const mapDispatchToProps = (dispatch: (action: Object) => void, { houseId }: { houseId: string }): Object => ({
  getHouse: () => dispatch({ type: 'GET_HOUSE', houseId }),
  openHouse: () => dispatch({ type: 'OPEN_HOUSE', houseId })
})
export default connect(mapStateToProps, mapDispatchToProps)(HouseItem)
