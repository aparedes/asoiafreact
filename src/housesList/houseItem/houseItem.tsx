/* @flow */
import './houseItem.css';
import React, { Component, MouseEvent } from 'react';
import { connect } from 'react-redux';
import { Dispatch, ReduxState } from '../../store/constants/reduxTypes';
import { houseActions } from '../../store/reducers/housesReducer';

interface DTP {
  getHouse: () => void;
  // openHouse: () => void;
}
type Props = STP & DTP & ReduxProps;

type State = {
  showFull: boolean;
};

export class HouseItem extends Component<Props, State> {
  static displayName = 'House';
  state = { showFull: false };

  componentDidUpdate(prevProps: Props, prevState: State) {
    if (
      !prevState.showFull &&
      this.state.showFull &&
      !this.props.got &&
      !this.props.getting
    ) {
      this.props.getHouse();
    }
  }

  changeState = () => {
    this.setState({
      showFull: !this.state.showFull,
    });
  };

  openExternal = (evt: MouseEvent<HTMLDivElement, globalThis.MouseEvent>) => {
    evt.preventDefault();
    evt.stopPropagation();
    // this.props.openHouse();
  };

  renderDetails = () => {
    const { coatOfArms, words } = this.props;
    return (
      <div className={'houseDetails'}>
        <div className={'houseDetail'}>
          <b>{'Coat of Arms: '}</b>
          {coatOfArms.length > 0 ? coatOfArms : '?'}
        </div>
        <div className={'houseDetail'}>
          <b>{'Words: '}</b>
          {words.length > 0 ? words : '?'}
        </div>
        <div className={'houseOpen'} onClick={this.openExternal}>
          <i className={'fa fa-external-link'} aria-hidden={'true'} />
        </div>
      </div>
    );
  };

  render() {
    const { error, getting } = this.props;
    return (
      <div className={'house'} data-testid="house" onClick={this.changeState}>
        <div className={'houseName'}>{this.props.name}</div>
        {this.state.showFull
          ? getting
            ? 'Loading'
            : error
            ? error
            : this.renderDetails()
          : null}
      </div>
    );
  }
}

type ReduxProps = { houseId: string };

interface STP {
  coatOfArms: string;
  error?: string;
  getting: boolean;
  got: boolean;
  name: string;
  words: string;
}

function mapStateToProps(state: ReduxState, { houseId }: ReduxProps): STP {
  const house = state.houses.houses.find((house) => house.id === houseId);
  return {
    coatOfArms: house?.coatOfArms ?? '',
    error: house?.error,
    getting: house?.getting ?? false,
    got: !!house?.coatOfArms,
    name: house?.name ?? '',
    words: house?.words ?? '',
  };
}

function mapDispatchToProps(dispatch: Dispatch, { houseId }: ReduxProps) {
  return {
    getHouse: () => {
      console.log(houseId);
      dispatch(houseActions.getHouse(houseId));
    },
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(HouseItem);
