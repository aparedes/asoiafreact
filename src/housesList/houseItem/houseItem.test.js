import React from 'react';
import { HouseItem } from './houseItem';
import renderer from 'react-test-renderer';

// coatOfArms: string,
// error: ?string,
// getHouse: () => void,
// getting: boolean,
// got: boolean,
// name: string,
// words: string,

function GetHouse() {}
test('Displays a house', () => {
  const component = renderer.create(
    <HouseItem
      houseId={'1'}
      name={'House Stark of Winterfell'}
      words={'Winter is Comming'}
      coatOfArms={'Dire wolf'}
      getHouse={GetHouse}
    />
  );
  let houseItem = component.toJSON();
  expect(houseItem).toMatchSnapshot();

  houseItem.props.onClick();
  houseItem = component.toJSON();
  expect(houseItem).toMatchSnapshot();
});
