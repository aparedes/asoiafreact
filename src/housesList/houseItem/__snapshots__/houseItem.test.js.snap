exports[`test Displays a house 1`] = `
<div
  className="house"
  onClick={[Function]}>
  <div
    className="houseName">
    House Stark of Winterfell
  </div>
</div>
`;

exports[`test Displays a house 2`] = `
<div
  className="house"
  onClick={[Function]}>
  <div
    className="houseName">
    House Stark of Winterfell
  </div>
  <div
    className="houseDetails">
    <div
      className="houseDetail">
      <b>
        Coat of Arms: 
      </b>
      Dire wolf
    </div>
    <div
      className="houseDetail">
      <b>
        Words: 
      </b>
      Winter is Comming
    </div>
  </div>
</div>
`;
